File: JSONParser.ts
type JSONObject = { [key: string]: JSONValue };
type JSONArray = JSONValue[];
type JSONText = string;
type JSONNumber = number;
type JSONBool = boolean;
type JSONNull = null;
type JSONValue =
  | JSONObject
  | JSONArray
  | JSONText
  | JSONNumber
  | JSONBool
  | JSONNull;


function parse(text: string): JSONValue {
  const tokens = tokenize(text);
  const [obj, rest] = parseTokens(tokens);
  if (rest.length > 0) {
    throw Error(`Unexpected tokens: ${rest}`);
  }
  return obj;
}

function parseTokens(tokens: Token[]): [JSONValue, Token[]] {
  const [t, ...tx] = tokens;
  if (t instanceof TokenOpenBrace) return parseObject(tx);
  if (t instanceof TokenText) return [t.value, tx];
  if (t instanceof TokenNumber) return [t.value, tx];
  if (t instanceof TokenBoolean) return [t.value, tx];
  if (t instanceof TokenNull) return [null, tx];
  throw Error(`Unexpected token ${t}, ${tx}`);
}

function parseObject(tokens: Token[]): [JSONObject, Token[]] {
  function go(tokens: Token[], acc: JSONObject): [JSONObject, Token[]] {
    const [t, ...tx] = tokens;
    if (t instanceof TokenCloseBrace) return [acc, tx];
    if (t instanceof TokenComma) return go(tx, acc);
    const [t1, ...tx1] = tx;
    if (t instanceof TokenText && t1 instanceof TokenColon) {
      const key = t.value;
      const [value, tx2] = parseTokens(tx1);
      return go(tx2, { ...acc, [key]: value });
    }
    throw Error(`Unexpected token 「${t}」, ${tx}`);
  }
  return go(tokens, {});
}

function splitFirstRestLast(txt: string): [string, string, string] {
  const first = txt[0];
  const rest = txt.slice(1);
  const last = rest.at(-1) ?? "";
  return [first, rest.slice(0, -1), last];
}

function parseText(textToken: string): JSONText {
  const [_first, rest, _last] = splitFirstRestLast(textToken);
  assert(_first === `"`);
  assert(_last === `"`);
  return rest;
}


type Token =
  | TokenOpenBrace
  | TokenCloseBrace
  | TokenColon
  | TokenComma
  | TokenText
  | TokenNumber
  | TokenBoolean
  | TokenNull;

class TokenBase {
  toString() {
    if ((this as any).value != null) {
      return `${this.constructor.name} ${(this as any).value}`;
    }
    return this.constructor.name;
  }
}
class TokenOpenBrace extends TokenBase {}
class TokenCloseBrace extends TokenBase {}
class TokenColon extends TokenBase {}
class TokenComma extends TokenBase {}
class TokenText extends TokenBase { constructor(public value: string) { super(); } }
class TokenNumber extends TokenBase { constructor(public value: number) { super(); } }
class TokenBoolean extends TokenBase { constructor(public value: boolean) { super(); } }
class TokenNull extends TokenBase {}

function tokenize(input: string): Token[] {
  const [c, cs] = [input[0], input.slice(1)];
  if (input === "") return [];
  if (c === "{") return [new TokenOpenBrace(), ...tokenize(cs)];
  if (c === "}") return [new TokenCloseBrace(), ...tokenize(cs)];
  if (c === ":") return [new TokenColon(), ...tokenize(cs)];
  if (c === " " || c === "\n") return [...tokenize(cs)];
  if (c === ",") return [new TokenComma(), ...tokenize(cs)];
  if (c === '"') {
    const [before, after] = splitFirstExceptEscape('"', cs);
    return [new TokenText(before), ...tokenize(after)];
  }
  if (c.match(/[\-\.0-9]/)) {
    const [before, after] = splitUntilNotNumber(cs);
    return [new TokenNumber(parseFloat(c + before)), ...tokenize(after)];
  }
  if (c + cs.slice(0, 3) === "true") return [new TokenBoolean(true), ...tokenize(cs.slice(3))];
  if (c + cs.slice(0, 4) === "false") return [new TokenBoolean(false), ...tokenize(cs.slice(4))];
  if (c + cs.slice(0, 3) === "null") return [new TokenNull(), ...tokenize(cs.slice(3))];
  throw Error(`Unexpected token 「${c}」, ${cs}`);
}

function splitUntilNotNumber(txt: string): [string, string] {
  const [c, cs] = [txt[0], txt.slice(1)];
  if (c.match(/[\-\.0-9]/)) {
    const [before, after] = splitUntilNotNumber(cs);
    return [c + before, after];
  }
  return ["", txt];
}

function splitFirstExceptEscape(ch: string, txt: string): [string, string] {
  function go(txt: string) {
    if (txt === "") return ["", ""];
    const [c, cs] = [txt[0], txt.slice(1)];
    if (c === ch) return ["", cs];
    if (c === "\\") {
      const [before, after] = go(cs.slice(1));
      return [cs[0] + before, after];
    }
    const [before, after] = go(cs);
    return [c + before, after];
  }
  return go(txt);
}

const json = JSON.stringify({
  "foo ": "bar",
  hoge: `fu"ga`,
  a: -23.2376,
  xxxxxx: true,
  parent: {
    child: "value",
  },
  null: null,
});

const result = tokenize(json);

const expected = [
  new TokenOpenBrace(),
  new TokenText("foo "),
  new TokenColon(),
  new TokenText("bar"),
  new TokenComma(),
  new TokenText("hoge"),
  new TokenColon(),
  new TokenText(`fu"ga`),
  new TokenComma(),
  new TokenText("a"),
  new TokenColon(),
  new TokenNumber(-23.2376),
  new TokenComma(),
  new TokenText("xxxxxx"),
  new TokenColon(),
  new TokenBoolean(true),
  new TokenComma(),
  new TokenText("parent"),
  new TokenColon(),
  new TokenOpenBrace(),
  new TokenText("child"),
  new TokenColon(),
  new TokenText("value"),
  new TokenCloseBrace(),
  new TokenComma(),
  new TokenText("null"),
  new TokenColon(),
  new TokenNull(),
  new TokenCloseBrace(),
];
assertEquals(result, expected);

const result2 = parse(json);
console.log(result2);

function assert(x: boolean) {
  if (!x) {
    throw Error("Fail to assertion");
  }
}

function assertEquals(left: Token[], right: Token[]) {
  for (let i = 0; i < left.length; i++) {
    const currentL = left[i];
    const currentR = right[i];
    if (currentL.toString() === currentR.toString()) {
      continue;
    }
    throw Error(
      `left and right are not same. index: ${i}, L: 「${currentL}」, R: 「${currentR}」`,
    );
  }
}

File: src/JSONParser.hs
module JSONParser (parse, JSONValue(..), tokenize, splitFirstExceptEscape, splitUntilNotNumber, JSONToken(..)) where
import Data.Char (isDigit, isSpace)

data JSONValue
  = JSONString String
  | JSONNumber Double
  | JSONBool Bool
  | JSONNull
  | JSONArray [JSONValue]
  | JSONObject [(String, JSONValue)]
  deriving (Show, Eq)

parse :: String -> JSONValue
parse input = let tokens = tokenize input
                  (result, rest) = parseTokens tokens
               in case rest of
                  [] -> result
                  _ -> error "Unexpected token"

parseTokens :: [JSONToken] -> (JSONValue, [JSONToken])
parseTokens (JSONTokenOpenBrace:tx) = parseObject tx
parseTokens (JSONTokenText t:tx) = (JSONString t, tx)
parseTokens (JSONTokenNumber t:tx) = (JSONNumber t, tx)
parseTokens (JSONTokenBoolean t:tx) = (JSONBool t, tx)
parseTokens (JSONTokenOpenBracket:tx) = parseArray tx
parseTokens (t:_) = error ("Unexpected token " ++ show t)
parseTokens [] = (JSONNull, [])

parseObject :: [JSONToken] -> (JSONValue, [JSONToken])
parseObject tx = go tx []
  where
    go :: [JSONToken] -> [(String, JSONValue)] -> (JSONValue, [JSONToken])
    go (JSONTokenCloseBrace:tx1) acc = (JSONObject (reverse acc), tx1)
    go (JSONTokenComma:tx1) acc = go tx1 acc
    go ((JSONTokenText key):JSONTokenColon:tx1) acc = let (value, tx2) = parseTokens tx1
                                                      in go tx2 ((key, value):acc)
    go t tx1 = error ("Unexpected token in object" ++ show t ++ show tx1)

parseArray :: [JSONToken] -> (JSONValue, [JSONToken])
parseArray tokens = go tokens []
  where
    go :: [JSONToken] -> [JSONValue] -> (JSONValue, [JSONToken])
    go [] _ = error "Unexpected end of input"
    go (JSONTokenComma:tx2) acc = go tx2 acc
    go (JSONTokenCloseBracket:tx2) acc = (JSONArray (reverse acc), tx2)
    go tx acc = let (value, tx2) = parseTokens tx
                 in go tx2 (value:acc)

data JSONToken
  = JSONTokenOpenBrace -- {
  | JSONTokenCloseBrace -- }
  | JSONTokenColon
  | JSONTokenComma
  | JSONTokenText String
  | JSONTokenNumber Double
  | JSONTokenBoolean Bool
  | JSONTokenNull
  | JSONTokenOpenBracket -- [
  | JSONTokenCloseBracket -- ]
  deriving (Show, Eq)

tokenize :: String -> [JSONToken]
tokenize ('{':xs) = JSONTokenOpenBrace : tokenize xs
tokenize ('}':xs) = JSONTokenCloseBrace : tokenize xs
tokenize ('[':xs) = JSONTokenOpenBracket : tokenize xs
tokenize (']':xs) = JSONTokenCloseBracket : tokenize xs
tokenize (':':xs) = JSONTokenColon : tokenize xs
tokenize (',':xs) = JSONTokenComma : tokenize xs
tokenize ('"':xs) = let (before, after) = splitFirstExceptEscape '"' xs
                         in JSONTokenText before:tokenize after
tokenize ('n':'u':'l':'l':xs) = JSONTokenNull : tokenize xs
tokenize ('t':'r':'u':'e':xs) = JSONTokenBoolean True : tokenize xs
tokenize ('f':'a':'l':'s':'e':xs) = JSONTokenBoolean False : tokenize xs
tokenize (x:xs)
  | isSpace x = tokenize xs
  -- 数字トークンは数字かマイナスで始まる。ピリオド始まりはありえない。
  | isDigit x || x == '-' = let (before, after) = splitUntilNotNumber xs
                 in JSONTokenNumber (read (x:before) :: Double) : tokenize after
  | otherwise = error ("Unexpected character: " ++ [x])
tokenize [] = []

splitUntilNotNumber :: String -> (String, String)
splitUntilNotNumber = go
  where
    go "" = ("", "")
    go (c:cs)
      | isDigit c || c == '.' = let (before, after) = go cs
                                 in (c:before, after)
      | otherwise = ("", c:cs)

splitFirstExceptEscape :: Char -> String -> (String, String)
splitFirstExceptEscape ch = go
  where
    go "" = ("", "")
    go (c:cs)
      | c == ch  = ("", cs)
      | c == '\\' = let (before, after) = go (tail cs)
                     in (head cs:before, after)
      | otherwise = let (before, after) = go cs
                     in (c:before, after)
